### PR TITLE
receiver_creator: add rules support

### PR DESCRIPTION
**Pull request by jrcamp**
receiver_creator: add rules support

 This adds rules support with some limitations in TODO. This allows dynamic
 discovery to work end to end with k8s pods and container ports. Will work on
 adding documentation but rules look like:

 ```
 type.port && name == "http" && pod.labels["region"] == "west-2"`
 ```

 Alternative for type check is to do:

 `type == types.Port`

 Or maybe some other way. First is shorter but second way is a bit more
 obvious. Thoughts welcome.

### CODE REVIEW

File: extension/observer/k8sobserver/config.go [hunk 1]
```diff
@@ -33,4 +33,7 @@ type Config struct {
 	//
 	// Then set this value to ${K8S_NODE_NAME} in the configuration.
 	Node string `mapstructure:"node"`
+
+	// StatusAddr is the address to [address]:port to listen on for status endpoint.
```
Comment (tigrannajaryan) [inline extension/observer/k8sobserver/config.go:37]: What does "is the address to [address]:port" mean?

File: extension/observer/k8sobserver/extension.go [hunk 2]
```diff
@@ -30,23 +33,42 @@ type k8sObserver struct {
...
 	stop     chan struct{}
 	config   *Config
+	server   *http.Server
 }
 
 func (k *k8sObserver) Start(ctx context.Context, host component.Host) error {
 	go k.informer.Run(k.stop)
+
+	// Use custom status port until there's a way to export status via zpages or something else.
+	if k.config.StatusAddr != "" {
```
Comment (tigrannajaryan) [inline extension/observer/k8sobserver/extension.go:43]: This appears to be unrelated to the primary change (adding rules support). Can you move to a separate PR to make reviewing, merging, reverting and searching commit history easier?

File: receiver/receivercreator/observerhandler.go [hunk 3]
```diff
@@ -67,9 +67,18 @@ func (obs *observerHandler) OnAdd(added []observer.Endpoint) {
 
 	for _, e := range added {
 		for _, template := range obs.receiverTemplates {
-			if !ruleMatches(template.Rule, e) {
+			if matches, err := template.rule.eval(e); err != nil {
+				obs.logger.Error("failed matching rule", zap.String("rule", template.Rule), zap.Error(err))
```
Comment (tigrannajaryan) [inline receiver/receivercreator/observerhandler.go:71]: Can this happen frequently? If so it may flood the logs. Consider using a metric instead or use `Debug` log level: https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#logging

File: receiver/receivercreator/observerhandler.go [hunk 4]
```diff
@@ -67,9 +67,18 @@ func (obs *observerHandler) OnAdd(added []observer.Endpoint) {
...
+				continue
+			} else if !matches {
 				continue
 			}
+
+			obs.logger.Info("starting receiver",
```
Comment (tigrannajaryan) [inline receiver/receivercreator/observerhandler.go:77]: Should this be `Debug` log instead? I am not sure how often this happens. Is this typically hapening only when new pods are added? That is probably fine, it is not too often.

File: receiver/receivercreator/rules.go [hunk 5]
```diff
@@ -15,10 +15,79 @@
...
 
-func ruleMatches(rule string, e observer.Endpoint) bool {
-	// TODO
-	return rule == "enabled"
+// rule wraps expr rule for later evaluation.
+type rule struct {
+	program *vm.Program
+}
+
+var ruleRe = regexp.MustCompile("^type.(Pod|Port)")
```
Comment (tigrannajaryan) [inline receiver/receivercreator/rules.go:33]: What does this specify?

File: receiver/receivercreator/rules.go [hunk 6]
```diff
@@ -15,10 +15,79 @@
...
+			"name":   o.Name,
+			"labels": o.Labels,
+		}
+	case observer.Port:
+		ruleTypes.Port = true
+		env = map[string]interface{}{
+			"type":     ruleTypes,
+			"name":     o.Name,
+			"port":     o.Port,
+			"pod_name": o.Pod.Name,
```
Comment (tigrannajaryan) [inline receiver/receivercreator/rules.go:77]: Should we use "pod.name" to match Otel conventions? Or it is not allowed as a variable name in `expr`?

Comment (tigrannajaryan) [reply receiver/receivercreator/rules.go:77]: This seems incorrect to me? Should we populate pod_name and pod labels if endpoint is `observer.Pod` instead? If this code is correct then it requires explanation since I do not understand it.

File: receiver/receivercreator/observerhandler.go (hunk 4 above)
Comment (jrcamp) [reply receiver/receivercreator/observerhandler.go:77]: Only when a pod is added that matches a rule and a receiver is started in response.

File: extension/observer/k8sobserver/config.go (hunk 1 above)
Comment (jrcamp) [reply extension/observer/k8sobserver/config.go:37]: That address is optional. (Removed from PR.)

File: receiver/receivercreator/observerhandler.go (hunk 3 above)
Comment (jrcamp) [reply receiver/receivercreator/observerhandler.go:71]: Possibly. It'll happen once per discovered pod per invalid rule. For example if a user writes a rule that uses a variable that doesn't exist it'll be logged per-pod. However it'll only logged be the first time it's observed. The same pod/rule won't be continuously logged. Do you think that's still too much?

File: receiver/receivercreator/rules.go (hunk 5 above)
Comment (jrcamp) [reply receiver/receivercreator/rules.go:33]: Comment added. It's a hacky regex for seeing if the rule starts with a type check.

File: receiver/receivercreator/rules.go (hunk 6 above)
Comment (jrcamp) [reply receiver/receivercreator/rules.go:77]: Yeah you can't use `.` in a variable name but I can still change it so that `pod.name` works. 

The reason labels and pod_name are there is because you may want to use them in your port check. For example:

`type.Port && name == "http" && pod.labels["app"] == "http"`

File: receiver/receivercreator/observerhandler.go (hunk 4 above)
Comment (tigrannajaryan) [reply receiver/receivercreator/observerhandler.go:77]: OK, I think it is infrequent enough. Should be fine to log it.

File: receiver/receivercreator/observerhandler.go (hunk 3 above)
Comment (tigrannajaryan) [reply receiver/receivercreator/observerhandler.go:71]: I think it is fine if it is once after the pod is observed. Let's keep it.

File: receiver/receivercreator/rules.go (hunk 6 above)
Comment (tigrannajaryan) [reply receiver/receivercreator/rules.go]: It would be great to use a dot in "pod.name". I think it will make it consistent with "pod.labels".

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, except status.go seems unnecessary.

File: extension/observer/status.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
+// Copyright 2020, OpenTelemetry Authors
```
[the remark this answers is not in the document]
Comment (jrcamp) [reply extension/observer/status.go:1]: Ah yeah, missed it. Thanks.

File: receiver/receivercreator/rules.go (hunk 6 above)
Comment (jrcamp) [reply receiver/receivercreator/rules.go]: Yeah there's `pod.name` and `pod.labels` now. There's still `name` which is the name of the port.

**Review by tigrannajaryan — APPROVED** [review]
